src/htr2hpc/api_client.py

```diff
@@ -0,0 +1,329 @@
+    if isinstance(data, dict):
+        # if a namedtuple already exists for this name, use it
+        nt_class = RESULTCLASS_REGISTRY.get(name)
+        # otherwise, create it and add it to the registry
+        if nt_class is None:
+            nt_class = namedtuple(name, data)
+            RESULTCLASS_REGISTRY[name] = nt_class
+        # once we have the class, initialize an instance with the given data
+        return nt_class(
+            # convert any nested objects to namedtuple classes
+            # use key name as namedtuple class name; convert plurals to singular
+            # NOTE: could add aliases other cleanup, e.g. shared_with_user(s) == user
+            **{key: to_namedtuple(key.rstrip("s"), val) for key, val in data.items()}
+        )
+
+    # when data is a list, convert each item in the list
+    # to a namedtuple with the specified name
+    elif isinstance(data, list):
+        return [to_namedtuple(name, d) for d in data]
+
+    # for all other data types, return as-is
+    else:
+        return data
+
+
+class eScriptoriumAPIClient:
+    def __init__(self, base_url: str, api_token: str):
+        # ensure no trailing slash before combining with other urls
+        self.base_url = base_url.rstrip("/")
+        self.api_root = f"{self.base_url}/api"
+        self.api_token = api_token
+        # create a request session, for request pooling
+        self.session = requests.Session()
+        headers = {
+            # set a user-agent header, but  preserve requests version information
+            "User-Agent": f"htr2hpc/{_version} ({self.session.headers['User-Agent']})",
+            "Authorization": f"Token {self.api_token}",
+        }
+        self.session.headers.update(headers)
+
+    def _make_request(
+        self,
+        url: str,
+        params: dict = None,
+        data: dict = None,
+        files: dict = None,
+        method: str = "GET",
+    ):
+        """
+        Make a GET request with the configured session. Takes a url
+        relative to :attr:`api_root` and optional dictionary of parameters for the request.
+        """
+        rqst_url = f"{self.api_root}/{url}"
+        rqst_opts = {}
+        if params:
+            rqst_opts["params"] = params.copy()
+
+        if method == "GET":
+            session_request = self.session.get
+        elif method in ["POST", "PUT"]:
+            session_request = getattr(self.session, method.lower())
+            # add post data and files to the request if any are specified
+            if data:
+                rqst_opts["data"] = data
+            if files:
+                rqst_opts["files"] = files
+        else:
+            raise ValueError(f"unsupported http method: {method}")
+
+        resp = session_request(rqst_url, **rqst_opts)
+        logger.debug(
+            "get %s %s: %f sec",
+            rqst_url,
+            resp.status_code,
+            resp.elapsed.total_seconds(),
```

Comment:
Perhaps use f-string
```suggestion
            f"get {rqst_url} {resp.status_code}: {resp.elapsed.total_seconds()} sec"
```